lib/common/common/src/iterator_ext/stoppable_iter.rs

```diff
@@ -0,0 +1,28 @@
+use std::sync::atomic::{AtomicBool, Ordering};
+
+pub struct StoppableIter<'a, I> {
+    iter: I,
+    is_stopped: &'a AtomicBool,
+}
+
+impl<'a, I> StoppableIter<'a, I> {
+    pub fn new(iter: I, is_stopped: &'a AtomicBool) -> Self {
+        Self { iter, is_stopped }
+    }
+}
+
+impl<'a, I> Iterator for StoppableIter<'a, I>
+where
+    I: Iterator,
+{
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // AI says, that if atomic is not updated frequently,
+        // it is cheap enough to check it on every iteration.
+        if self.is_stopped.load(Ordering::Relaxed) {
+            return None;
+        }
```

Comment:
So here's a stupid idea:

I'm curious if we could just ignore atomics here and concurrently set the boolean.

It'd be cheaper by not using atomic operations, at the cost of potentially 'recognizing' the changed flag a bit later due to local caches.